### PR TITLE
Handle stale messages 263

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: p2p/protocol.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	pendMutex          sync.RWMutex
+	pendingMap         map[uint64]chan interface{}
+	pendingQueue       *list.List
+	resHandlers        map[uint64]func(msg []byte)
+	msgRequestHandlers map[MessageType]func(msg []byte) []byte
+	ingressChannel     chan service.Message
+	requestLifetime    time.Duration
+}
+
+func NewProtocol(network Service, name string, ttl time.Duration) *Protocol {
```
Comment (zalmen) [inline p2p/protocol.go]: document the method, not sure if the purpose of 'ttl' is clear to the user of this class

File: p2p/protocol.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+			//todo add buffer and option to limit number of concurrent goroutines
+			go p.handleMessage(msg)
+		case <-timer.C:
+			go p.cleanStaleMessages()
+		}
+	}
+}
+
+func (p *Protocol) cleanStaleMessages() {
+	for {
```
Comment (zalmen) [inline p2p/protocol.go:70]: you should run this under a writer lock

Comment (zalmen) [reply p2p/protocol.go:70]: Also, what about cleaning the resHandlers map?

**Review by zalmen — DISMISSED** [review]

File: p2p/protocol.go (hunk 2 above)
Comment (almogdepaz) [reply p2p/protocol.go:70]: good idea, done.

File: p2p/protocol.go (hunk 1 above)
Comment (almogdepaz) [reply p2p/protocol.go]: added documentation to the struct and change the var name in the constructor

File: p2p/protocol.go [hunk 3]
```diff
@@ -84,13 +113,13 @@ func (p *Protocol) handleRequestMessage(sender crypto.PublicKey, headers *pb.Mes
 
 func (p *Protocol) handleResponseMessage(headers *pb.MessageWrapper) {
 
-	//get and remove from pending
-	p.pendMutex.Lock()
-	pend, okPend := p.pending[headers.ReqID]
+	//get and remove from pendingMap
+	p.pendMutex.RLock()
```
Comment (y0sher) [inline p2p/protocol.go:117]: you need `Lock` because you delete

File: p2p/protocol.go [hunk 4]
```diff
@@ -84,13 +113,13 @@ func (p *Protocol) handleRequestMessage(sender crypto.PublicKey, headers *pb.Mes
...
+	pend, okPend := p.pendingMap[headers.ReqID]
 	foo, okFoo := p.resHandlers[headers.ReqID]
-	delete(p.pending, headers.ReqID)
+	delete(p.pendingMap, headers.ReqID)
 	delete(p.resHandlers, headers.ReqID)
-	p.pendMutex.Unlock()
+	p.pendMutex.RUnlock()
```
Comment (y0sher) [inline p2p/protocol.go:122]: Unlock

**Review by y0sher — APPROVED** [review]

**Pull request merged**